File: Notes/Computer Science/Algorithms/Meta/Sebastien Rubens Meta Puzzles Solutions/scala/l3_slippery_strip.scala
// meta_puzzles by Sebastien Rubens
//
// Please go to https://github.com/seb-pg/meta_puzzles/README.md
// for more information
//
// To the extent possible under law, the person who associated CC0 with
// meta_puzzles has waived all copyright and related or neighboring rights
// to meta_puzzles.
//
// You should have received a copy of the CC0 legalcode along with this
// work.  If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.

package l3_slippery_strip

import scala.collection.mutable.ArrayBuffer

class Counts(var star: Int,
             var right: Int,
             var down: Int) {
}

object Solution {
    def getMaxCollectableCoins(R: Int, C: Int, G: Array[Array[String]]): Int = {

        type char_counter_t = ArrayBuffer[Int]

        def get_counts(row: Array[Char], counts: char_counter_t): Counts = {
            counts('*') = 0
            counts('>') = 0
            counts('v') = 0
            for (c <- row) {
                var d = c
                if (d >= counts.size)
                    d = '.' // empty value (we do not use it)
                counts(d) += 1
            }
            return new Counts(counts('*'), counts('>'), counts('v'))
        }

        def get_nb_coins_right_then_down3(row: Array[Char], _count_down: Int, _count_right: Int): Int = {
            // complexity : O(C)
            if (_count_down == 0)
                return 0
            val k = row.indexOf('v')
            val j = if (k != -1) (k + 1) else 0
            val new_row = new ArrayBuffer[Char]()
            new_row.addAll(row.slice(j, row.size))
            new_row.addAll(row.slice(0, j))
            var count_down = _count_down
            var count_right = _count_right
            var nb_coins_right_then_down = 0
            val size = new_row.size
            var last = 0
            while (count_right * count_down != 0) {
                val first = new_row.slice(last, size).indexOf('>') + last
                last = new_row.slice(first, size).indexOf('v') + 1 + first
                nb_coins_right_then_down = nb_coins_right_then_down.max(new_row.slice(first, last).count((c: Char) => c == '*'))
                count_down -= 1
                count_right -= new_row.slice(first, last).count((c: Char) => c == '>')
            }
            return nb_coins_right_then_down
        }

        def _getMaxCollectableCoins(R: Int, C: Int, G: Array[Array[Char]]): Int = {
            val counts = ArrayBuffer.fill(256)(0)
            var res = 0
            for (i <- 0 until G.size) {
                val row = G(G.size - i - 1)
                val count = get_counts(row, counts)
                val nb_coins_immediately_down = count.star.min(1)
                if (count.right == C)
                    res = 0
                else if (count.right == 0)
                    res += nb_coins_immediately_down
                else {
                    val nb_coins_right_then_down = get_nb_coins_right_then_down3(row, count.down, count.right)
                    val nb_coins_right_forever = if (count.down == 0) count.star else 0
                    res = (nb_coins_immediately_down + res).max((nb_coins_right_then_down + res).max(nb_coins_right_forever))
                }
            }
            return res.toInt // result should always be positive
        }

        val H = new ArrayBuffer[Array[Char]](G.size)
        for (in_row <- G)
        {
            val row = new ArrayBuffer[Char](in_row.size)
            for (c <- in_row)
                row += c(0)
            H += row.toArray
        }
        return _getMaxCollectableCoins(H.size, if (H.isEmpty) 0 else H(0).size, H.toArray)
    }


    class Args(val G: Array[String],
               val res: Int ) extends test.Result[Int] {
        override def get_result(): Int = res
    }

    def tests(): Int =
    {
        def _getMaxCollectableCoinsTest(G: Array[String]): Int = {
            val H = new ArrayBuffer[Array[String]](G.size)
            for (in_row <- G)
            {
                val row = new ArrayBuffer[String](in_row.length)
                for (c <- in_row)
                    row += c.toString()
                H += row.toArray
            }
            return getMaxCollectableCoins(H.size, if (H.isEmpty) 0 else H(0).size, H.toArray)
        }


        val wrapper = (p: Args) => _getMaxCollectableCoinsTest(p.G)

        val args_list = Array[Args](
            new Args(Array[String](".***", "**v>", ".*.."), 4),
            new Args(Array[String](">**", "*>*", "**>"), 4),
            new Args(Array[String](">>", "**"), 0),
            new Args(Array[String](">*v*>*", "*v*v>*", ".*>..*", ".*..*v"), 6),
            // extra1
            new Args(Array[String](), 0),
            new Args(Array[String]("."), 0),
            new Args(Array[String]("v"), 0),
            new Args(Array[String](">"), 0),
            new Args(Array[String]("*"), 1),
            // extra2
            new Args(Array[String](".", ".", ">", "*"), 0),
            new Args(Array[String](".", "*", ">", "*"), 1),
            new Args(Array[String](".", "*", ">", "."), 1),
            new Args(Array[String]("*", ".", ">", "."), 1),
            new Args(Array[String]("***", "...", ">vv", "..."), 1),
            // extra3
            new Args(Array[String]("*....", ".*...", "..*..", "...*."), 4),
            new Args(Array[String]("....", "....", "....", "...."), 0),
            new Args(Array[String]("***>", "...."), 3),
            new Args(Array[String]("*"), 1),
            new Args(Array[String]("vvvv"), 0),
            new Args(Array[String]("vvvv", "....", ">>>>"), 0),
            // extra4
            new Args(Array[String]("******", "......", ">*>vvv", "......"), 2),
            new Args(Array[String]("*****", ".....", ">>vvv", "....."), 1)
        )

        return test.TestAll.run_all_tests("l3_slippery_strip", args_list, wrapper)
    }
}

File: Notes/Computer Science/Algorithms/Meta/Sebastien Rubens Meta Puzzles Solutions/kotlin/l3_slippery_strip.kt
// meta_puzzles by Sebastien Rubens
//
// Please go to https://github.com/seb-pg/meta_puzzles/README.md
// for more information
//
// To the extent possible under law, the person who associated CC0 with
// meta_puzzles has waived all copyright and related or neighboring rights
// to meta_puzzles.
//
// You should have received a copy of the CC0 legalcode along with this
// work.  If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.

package l3_slippery_strip

class Counts(
    var star: UInt,
    var right: UInt,
    var down: UInt, )

typealias row_t = ArrayList<Char>;
typealias char_counter_t = ArrayList<UInt>;

fun get_counts(row: Array<Char>, counts: char_counter_t): Counts
{
    counts['*'.code] = 0U;
    counts['>'.code] = 0U;
    counts['v'.code] = 0U;
    for (c in row)
    {
        var d = c.code;
        if (d >= counts.size)
            d = '.'.code;  // empty value (we do not use it)
        ++counts[d];
    }
    return Counts(counts['*'.code], counts['>'.code], counts['v'.code])
}

fun get_nb_coins_right_then_down3(row: Array<Char>, _count_down: UInt, _count_right: UInt): UInt
{
    // complexity : O(C)
    if (_count_down == 0U)
        return 0U;
    val k = row.indexOf('v');
    val j = if (k != -1) (k + 1) else 0;
    val new_row = row_t();
    new_row.addAll(row.slice(j until row.size));
    new_row.addAll(new_row.size, row.slice(0 until j));
    var count_down = _count_down;
    var count_right = _count_right;
    var nb_coins_right_then_down = 0U;
    val size = new_row.size;
    var last = 0;
    while (count_right * count_down != 0U)
    {
        val first = new_row.slice(last until size).indexOf('>') + last;
        last = new_row.slice(first until size).indexOf('v') + 1 + first;
        nb_coins_right_then_down = maxOf(nb_coins_right_then_down, new_row.slice(first until last).count() {it == '*'}.toUInt())
        count_down -= 1U;
        count_right -= new_row.slice(first until last).count() { it == '>' }.toUInt();
    }
    return nb_coins_right_then_down;
}

fun _getMaxCollectableCoins(R: Int, C: Int, G: Array<Array<Char>>): Int {
    val counts = char_counter_t(256);
    repeat(256) { index -> counts.add(0U) }
    var res = 0U;
    for (i in 0 until G.size)
    {
        val row = G[G.size - i - 1];
        val count = get_counts(row, counts);
        val nb_coins_immediately_down = minOf(count.star, 1u);
        if (count.right == C.toUInt())
            res = 0U;
        else if (count.right == 0U)
            res += nb_coins_immediately_down;
        else
        {
            val nb_coins_right_then_down = get_nb_coins_right_then_down3(row, count.down, count.right);
            val nb_coins_right_forever = if (count.down == 0U) count.star else 0U;
            res = maxOf(nb_coins_immediately_down + res, maxOf(nb_coins_right_then_down + res, nb_coins_right_forever));
        }
    }
    return res.toInt();  // result should always be positive
}

fun getMaxCollectableCoins(R: Int, C: Int, G: Array<Array<String>>): Int {
    val H = ArrayList<Array<Char>>(G.size);
    for (in_row in G)
    {
        val row = ArrayList<Char>(in_row.size);
        for (c in in_row)
            row.add(if (!c.isEmpty()) c[0] else '.');  // FIXME: Meta's website provide corrupted test data!
        H.add(row.toTypedArray());
    }
    return _getMaxCollectableCoins(H.size, if (H.isEmpty()) 0 else H[0].size, H.toTypedArray());
}

fun _getMaxCollectableCoinsTest(G: Array<String>): Int {
    val H = ArrayList<Array<String>>(G.size);
    for (in_row in G)
    {
        val row = ArrayList<String>(in_row.length);
        for (c in in_row)
            row.add(c.toString());
        H.add(row.toTypedArray());
    }
    return getMaxCollectableCoins(H.size, if (H.isEmpty()) 0 else H[0].size, H.toTypedArray());
}

class Args(
    val G: Array<String>,
    val res: Int, ) : test.Result<Int> {
    override fun get_result(): Int { return res; };
}

fun tests(): UInt
{
    val wrapper = { p: Args -> _getMaxCollectableCoinsTest(p.G) };

    val args_list: List<Args> = listOf(
        Args( arrayOf(".***", "**v>", ".*.."), 4 ),
        Args( arrayOf(">**", "*>*", "**>"), 4 ),
        Args( arrayOf(">>", "**"), 0 ),
        Args( arrayOf(">*v*>*", "*v*v>*", ".*>..*", ".*..*v"), 6 ),
        // extra1
        Args( arrayOf<String>(), 0 ),
        Args( arrayOf("."), 0 ),
        Args( arrayOf("v"), 0 ),
        Args( arrayOf(">"), 0 ),
        Args( arrayOf("*"), 1 ),
        // extra2
        Args( arrayOf(".", ".", ">", "*"), 0 ),
        Args( arrayOf(".", "*", ">", "*"), 1 ),
        Args( arrayOf(".", "*", ">", "."), 1 ),
        Args( arrayOf("*", ".", ">", "."), 1 ),
        Args( arrayOf("***", "...", ">vv", "..."), 1 ),
        // extra3
        Args( arrayOf("*....", ".*...", "..*..", "...*."), 4 ),
        Args( arrayOf("....", "....", "....", "...."), 0 ),
        Args( arrayOf("***>", "...."), 3 ),
        Args( arrayOf("*"), 1 ),
        Args( arrayOf("vvvv"), 0 ),
        Args( arrayOf("vvvv", "....", ">>>>"), 0 ),
        // extra4
        Args( arrayOf("******", "......", ">*>vvv", "......"), 2 ),
        Args( arrayOf("*****", ".....", ">>vvv", "....."), 1 ),
    );

    return test.run_all_tests("l3_slippery_strip", args_list, wrapper);
}
